apps/lib/apps.c

```diff
@@ -581,31 +581,35 @@ EVP_PKEY *load_key(const char *uri, int format, int may_stdin,
     if (desc == NULL)
         desc = "private key";
 
-    if (format == FORMAT_ENGINE) {
+    if (format == FORMAT_ENGINE)
         uri = allocated_uri = make_engine_uri(e, uri, desc);
-    }
     (void)load_key_certs_crls(uri, format, may_stdin, pass, desc,
                               &pkey, NULL, NULL, NULL, NULL, NULL, NULL);
 
     OPENSSL_free(allocated_uri);
     return pkey;
 }
 
+/* first try reading public key, on failure resort to loading private key */
 EVP_PKEY *load_pubkey(const char *uri, int format, int maybe_stdin,
                       const char *pass, ENGINE *e, const char *desc)
 {
     EVP_PKEY *pkey = NULL;
     char *allocated_uri = NULL;
+    BIO *bio_bak = bio_err;
 
     if (desc == NULL)
         desc = "public key";
 
-    if (format == FORMAT_ENGINE) {
+    if (format == FORMAT_ENGINE)
         uri = allocated_uri = make_engine_uri(e, uri, desc);
-    }
+    bio_err = NULL;
     (void)load_key_certs_crls(uri, format, maybe_stdin, pass, desc,
                               NULL, &pkey, NULL, NULL, NULL, NULL, NULL);
-
+    bio_err = bio_bak;
+    if (pkey == NULL)
+        (void)load_key_certs_crls(uri, format, maybe_stdin, pass, desc,
+                                  &pkey, NULL, NULL, NULL, NULL, NULL, NULL);
```

Comment:
Is the error stack relevant here?  If the public key and private load fails, what will be reported?

Another way to handle this would be:

* Add another argument to `load_key_certs_crls` to request *quiet* operation, with no error output and the error stack cleared.
* Try to load a private key first, with `quiet=1`
* Finally, if that fails, try load a public key with `quiet=0`.

This avoids the kludgey messing around with bio_err...

Reply:
Good point - done.

On this occasion spotted and removed a leftover outdated decl in `apps.h`:
```
int load_key_cert_crl(const char *uri, int format, int maybe_stdin,
                      const char *pass, const char *desc,
                      EVP_PKEY **ppkey, EVP_PKEY **ppubkey,
                      X509 **pcert, X509_CRL **pcrl);
```
and made `check-format.pl` happy on `apps/include/apps.h`.